Use c10::variant-based enums for Reduction

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28806 Fix implementation of F::kl_div / F::mse_loss / F::binary_cross_entropy
* **#27942 Use c10::variant-based enums for Reduction**

Differential Revision: [D18202857](https://our.internmc.facebook.com/intern/diff/D18202857)

Commit:
Use c10::variant-based enums for Reduction

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

CarMiranda commented on torch/csrc/api/include/torch/nn/options/loss.h:
```diff
@@ -90,7 +92,7 @@ struct TORCH_API TripletMarginLossOptions {
   TORCH_ARG(bool, swap) = false;
   /// Specifies the reduction to apply to the output. Default: Mean
-  TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean;
+  TORCH_ARG(10::variant<enumtype::kNone, enumtype::kMean, enumtype::kSum>, reduction) = torch::kMean;
```
Sorry I was looking if a PR of mine was in the list and found this PR which kind of concerns what I was doing :P I think you are missing a `c` here (should be `c10`) and in previous declarations too!

yf225 replied:
@CarMiranda Thanks a lot for the catch! I just fixed it. After this PR is merged, I will do a sweep to change all `torch::nn` layers that use `Reduction` to use the corresponding variant type. :D

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction, and fix F::kl_div / mse_loss / binary_cross_entropy"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

yf225 commented on torch/csrc/api/include/torch/enum.h:
```diff
@@ -26,10 +29,21 @@ const enumtype::k##name k##name; \
+OPTIONS_NAME(torch::enumtype::TYPE1 ARG_NAME) : ARG_NAME##_(torch::TYPE1) {} \
+OPTIONS_NAME(torch::enumtype::TYPE2 ARG_NAME) : ARG_NAME##_(torch::TYPE2) {} \
+OPTIONS_NAME(torch::enumtype::TYPE3 ARG_NAME) : ARG_NAME##_(torch::TYPE3) {} \
+OPTIONS_NAME(torch::enumtype::TYPE4 ARG_NAME) : ARG_NAME##_(torch::TYPE4) {}
```
TODO: I need to add a comment for why we need these two macros

ezyang replied:
Yes plz lol

yf225 replied:
I added the comment :D

Commit:
Update on "Use c10::variant-based enums for Reduction, and fix F::kl_div / mse_loss / binary_cross_entropy"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

ezyang commented on torch/csrc/api/include/torch/enum.h:
````diff
@@ -26,10 +29,69 @@ const enumtype::k##name k##name; \
+// ```
+// Tensor some_functional(
+//     const Tensor& input,
+//     const SomeOptions& options = {}) {
````
This smells wrong to me. Why don't you just take it by value? It's a very small struct.

yf225 replied:
I updated the comment here and will do a sweep to take `Options` by value in all functionals in a follow-up PR. Although it doesn't fix the problem that `TORCH_OPTIONS_CTOR_VARIANT_ARG3`/`TORCH_OPTIONS_CTOR_VARIANT_ARG4` try to address though :(

ezyang commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -10,37 +11,104 @@ inline Tensor l1_loss(
+  return torch::l1_loss(
+    input,
+    target,
+    c10::visit(enumtype::_reduction_get_enum{}, options.reduction()));
```
Instead of manually typing out `c10::visit` everywhere, why not come up with a good API for doing this and call that instead? Especially since `_reduction_get_enum` is underscored...

yf225 replied:
Thanks for the suggestion! I added `torch::enumtype::reduction_get_enum()` API for this purpose :D

ezyang commented:
Can you say more about what "fix F::kl_div / mse_loss / binary_cross_entropy" means?

Commit:
Update on "Use c10::variant-based enums for Reduction, and fix F::kl_div / mse_loss / binary_cross_entropy"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction, and fix F::kl_div / mse_loss / binary_cross_entropy"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction, and fix F::kl_div / mse_loss / binary_cross_entropy"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head

yf225 commented:
> Can you say more about what "fix F::kl_div / mse_loss / binary_cross_entropy" means?

The original logic in F::kl_div / mse_loss / binary_cross_entropy doesn't match that of Python version. I moved the changes to another PR since it is not strictly related to the Reduction enum changes.

Commit:
Update on "Use c10::variant-based enums for Reduction"

Use c10::variant-based enums for Reduction

gh-metadata: pytorch pytorch 27942 gh/yf225/11/head
